onnxscript/rewriter/pattern.py

```diff
@@ -365,18 +389,56 @@ class NodePattern:
 
     def __init__(
         self,
-        domain: Pattern[str],
-        op: Pattern[str],
+        domain: OpsetPatternBuilder,
+        op: str | Pattern[str],
         inputs: Sequence[int | float | ValuePattern | None],
         attributes: dict[str, AttrPattern],
+        outputs: Sequence[str],
     ):
         self.domain = domain
-        self.op = op
+        self.op = StringConstantPattern(op) if isinstance(op, str) else op
         self.inputs = [_to_value_pattern(x) for x in inputs]
         self.attributes = attributes
+        # In the common case, domain and op are constants, which can be used to optimize matching.
+        if isinstance(op, str) and domain.domain_name is not None:
+            # TODO(rama): support overloaded operators.
+            overload = ""
+            self._op_identifier = (domain.domain_name, op, overload)
+        else:
+            self._op_identifier = None
+        self.outputs = [NodeOutputPattern(self, i, name) for i, name in enumerate(outputs)]
 
-    def matches_node(self, node: ir.Node) -> MatchResult:
-        """Examine if the IR node matches the self pattern."""
+    def op_identifier(self) -> Tuple[str, str, str]:
+        return self._op_identifier
+
+    @property
+    def op_type(self) -> str:
+        if self._op_identifier is not None:
+            return self._op_identifier[1]
```

Comment:
Should we just return self.op?

Reply:
No ... remember that we can even specify that a pattern-node can match any node whose op has "Mat" as a prefix. So, the pattern's op is a predicate (an object with a `match` method). The op-identifier has the opname, when it is available.

Reply:
I still don't get it. I thought domain is the only arg that we support for prefix? Could you share me the example pattern/test?

Reply:
The example is the one using [submodule](https://github.com/microsoft/onnxscript/blob/b4dd7774f166e46984aebc6d3626cd5edae3dcd5/onnxscript/rewriter/onnxruntime/group_normalization_merge_silu.py#L31) ... I think you added it?

Reply:
I see. I forgot op_type is PrefixPattern. I thought this was the same op_type we have in ir.Node.